gpioioctl/gpio.go

```diff
@@ -0,0 +1,617 @@
+func (chip *GPIOChip) String() string {
+	s := fmt.Sprintf("{\"Name\": \"%s\", \"Path\": \"%s\", \"Label\": \"%s\", \"LineCount\": %d, \"Lines\": [ \n",
+		escapeJSONString(chip.Name()), escapeJSONString(chip.Path()), escapeJSONString(chip.Label()), chip.LineCount())
+	for _, line := range chip.lines {
+		s += line.String() + ",\n"
+	}
+	s = s[:len(s)-2] + "],"
+	s += "\n\"LineSets\": [ \n"
+	for _, ls := range chip.lineSets {
+		if ls.fd > 0 {
+			s += ls.String() + ",\n"
+		}
+	}
+	s = s[:len(s)-2] + "]}"
+	return s
+}
+
+// LineSet requests a set of io pins and configures them according to the
+// parameters. Using a LineSet, you can perform IO operations on multiple
+// lines in a single operation. For more control, see LineSetFromConfig.
+func (chip *GPIOChip) LineSet(defaultDirection LineDir, defaultEdge gpio.Edge, defaultPull gpio.Pull, lines ...string) (*LineSet, error) {
+	cfg := &LineSetConfig{DefaultDirection: defaultDirection, DefaultEdge: defaultEdge, DefaultPull: defaultPull}
+	for _, lineName := range lines {
+		p := chip.ByName(lineName)
+		if p == nil {
+			return nil, fmt.Errorf("line %s not found", lineName)
+		}
+		cfg.Lines = append(cfg.Lines, p.Name())
+	}
+	return chip.LineSetFromConfig(cfg)
+}
+
+// driverGPIO implements periph.Driver.
+type driverGPIO struct {
+	_ string
+}
+
+func (d *driverGPIO) String() string {
+	return "ioctl-gpio"
+}
+
+func (d *driverGPIO) Prerequisites() []string {
+	return nil
+}
+
+func (d *driverGPIO) After() []string {
+	return nil
+}
+
+// Init initializes GPIO ioctl handling code.
+//
+// # Uses Linux gpio ioctl as described at
+//
+// https://docs.kernel.org/userspace-api/gpio/chardev.html
+func (d *driverGPIO) Init() (bool, error) {
+	items, err := filepath.Glob("/dev/gpiochip*")
+	if err != nil {
+		return true, err
+	}
+	if len(items) == 0 {
+		return false, errors.New("no GPIO chips found")
+	}
+	Chips = make([]*GPIOChip, 0)
+	for _, item := range items {
+		chip, err := newGPIOChip(item)
+		if err != nil {
+			log.Println("gpioioctl.driverGPIO.Init() Error", err)
+			return false, err
+		}
+		Chips = append(Chips, chip)
+		for _, line := range chip.lines {
+			if len(line.name) > 0 && line.name != "_" && line.name != "-" {
+				if err = gpioreg.Register(line); err != nil {
+					log.Println("chip", chip.Name(), " gpioreg.Register(line) ", line, " returned ", err)
+				}
+			}
+		}
+	}
+	return true, nil
+}
+
+var drvGPIO driverGPIO
+
+func init() {
+	if runtime.GOOS == "linux" {
+
+		// Init our consumer name. It's used when a line is requested, and
+		// allows utility programs like gpioinfo to find out who has a line
+		// open.
+		fname := path.Base(os.Args[0])
+		s := fmt.Sprintf("%s@%d", fname, os.Getpid())
+		charBytes := []byte(s)
+		if len(charBytes) >= _GPIO_MAX_NAME_SIZE {
+			charBytes = charBytes[:_GPIO_MAX_NAME_SIZE-1]
+		}
+		consumer = charBytes
+
+		driverreg.MustRegister(&drvGPIO)
+	}
+}
+
+func escapeJSONString(str string) string {
```

Comment:
FYI, the better implementation is to use:
```
d, _ := json.Marshal(str)
return string(d)
```

https://cs.opensource.google/go/go/+/refs/tags/go1.23.1:src/encoding/json/encode.go;l=160 ultimately leads to https://cs.opensource.google/go/go/+/refs/tags/go1.23.1:src/encoding/json/encode.go;l=967 which has many subtleties.

That sad, I don't think you need this function given the other comment.